src/client/components/scroller/scroller.tsx

```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
 
 import * as React from 'react';
 import * as ReactDOM from 'react-dom';
+import { classNames, getXFromEvent, getYFromEvent } from '../../utils/dom/dom';
 import { Fn } from '../../../common/utils/general/general';
 
+export interface ScrollerLayout {
+  bodyWidth: number;
+  bodyHeight: number;
+
+  top: number;
+  right: number;
+  bottom: number;
+  left: number;
+};
+
 export interface ScrollerProps extends React.Props<any> {
-  onScroll: Fn;
-  style?: Lookup<any>;
-  ref?: string;
-  className?: string;
-  onMouseLeave?: Fn;
-  onMouseMove?: Fn;
-  onClick?: Fn;
+  layout: ScrollerLayout;
+
+  onClick?: (x: number, y: number) => void;
+  onMouseMove?: (x: number, y: number) => void;
+  onMouseLeave?: () => void;
+  onScroll?: (scrollTop: number, scrollLeft: number) => void;
+
+  // Transcluded elements
+  topGutter?: JSX.Element | JSX.Element[];
+  rightGutter?: JSX.Element | JSX.Element[];
+  bottomGutter?: JSX.Element | JSX.Element[];
+  leftGutter?: JSX.Element | JSX.Element[];
+  topLeftCorner?: JSX.Element | JSX.Element[];
+  topRightCorner?: JSX.Element | JSX.Element[];
+  bottomRightCorner?: JSX.Element | JSX.Element[];
+  bottomLeftCorner?: JSX.Element | JSX.Element[];
+  body?: JSX.Element[];
+  bodyOverlay?: JSX.Element | JSX.Element[];
 }
 
 export interface ScrollerState {
-  scrollLeft: number;
-  scrollTop: number;
+  scrollTop?: number;
+  scrollLeft?: number;
+
+  viewportHeight?: number;
+  viewportWidth?: number;
 }
 
 export class Scroller extends React.Component<ScrollerProps, ScrollerState> {
-
   constructor() {
     super();
     this.state = {
+      scrollTop: 0,
       scrollLeft: 0,
-      scrollTop: 0
+      viewportHeight: 0,
+      viewportWidth: 0
+    };
+  }
+
+  getTopLeftCornerStyle(): React.CSSProperties {
```

Comment:
I think it would be shorter and simpler to write these functions as:

```typescript
type XPos = 'left' | 'right';
type yPos = 'top' | 'bottom';

getCornerStyle(xPos: XPos, yPos: YPos): React.CSSProperties {
  const { layout } = this.props;
  var style = {};
  if (xPos === 'left') {
    style.left = 0;
    style.width = layout.left;
  } else {
    /* ... */
  }

  if (yPos === 'top') {
    /* ... */
  } else {
    /* ... */
  }
  return style;
}
```

Then you would not need kebab-case transformer function